### PR TITLE
Feature/elyard/controller implementation

### PR DESCRIPTION
**Pull request by SantoshKesani**
Please look at the updates and suggest changes if any.

### CODE REVIEW

File: app/Controller.cpp [hunk 1]
```diff
@@ -80,9 +85,16 @@ void Controller::controlLoop() {
...
-    // @TODO groundbreaking robotics goes here
-    //
+    double speed_error, heading_error, command_throttle, command_steering, actual_throttle, actual_steering;
+    double steeringVel = 0.0; // TODO: Understanding the current_steering_vel and update accordingly
+
+    this->model_->getError(speed_error, heading_error);
+    command_throttle = this->pid_speed_->getCommand(speed_error);
+    command_steering = this->pid_heading_->getCommand(heading_error);
+  
+    this->model_->limits_->limit(command_throttle, command_steering, steeringVel, actual_throttle, actual_steering, steeringVel, duration);
```
Comment (danielmohansahu) [inline app/Controller.cpp]: I believe (@selyard can confirm) the limiting is done internally by the `model` class under Spencer's new setup.

File: app/Controller.cpp [hunk 2]
```diff
@@ -56,18 +56,23 @@ bool Controller::isRunning() const {
...
 void Controller::setGoal(const double heading, const double speed) {
+  this->model_->setGoal(heading, speed);
 }
 
 void Controller::getGoal(double& heading, double& speed) const {
+  this->model_->getGoal(heading, speed);
 }
 
 void Controller::getCommand(double& throttle, double& steering) const {
+  this->model_->getCommand(heading, speed);
```
Comment (danielmohansahu) [inline app/Controller.cpp]: These are the wrong variable names.

File: app/Controller.cpp [hunk 3]
```diff
@@ -80,9 +85,16 @@ void Controller::controlLoop() {
     // update next target time
     next_loop_time += duration;
 
-    //
-    // @TODO groundbreaking robotics goes here
-    //
+    double speed_error, heading_error, command_throttle, command_steering, actual_throttle, actual_steering;
+    double steeringVel = 0.0; // TODO: Understanding the current_steering_vel and update accordingly
```
Comment (danielmohansahu) [inline app/Controller.cpp]: This needs to be implemented as part of the PR

File: app/Controller.cpp [hunk 4]
```diff
@@ -80,9 +85,16 @@ void Controller::controlLoop() {
...
+
+    this->model_->getError(speed_error, heading_error);
+    command_throttle = this->pid_speed_->getCommand(speed_error);
+    command_steering = this->pid_heading_->getCommand(heading_error);
+  
+    this->model_->limits_->limit(command_throttle, command_steering, steeringVel, actual_throttle, actual_steering, steeringVel, duration);
+  
+    this->model_->command(actual_throttle, actual_steering, duration);
```
Comment (danielmohansahu) [inline app/Controller.cpp]: This is the wrong duration variable; you'll need to convert it to a double to get this to run.

**Review by danielmohansahu — CHANGES_REQUESTED** [review]
This file causes the build to break. Before this is merged you need to:
- [x] Confirm that this builds locally on your machine (`mkdir build && cd build && cmake .. && make`)
~~- [ ] Check if all the `Controller` unit tests are passing (`./test/cpp-test` and look for the Controller section)~~

I'm not sure all of the `Controller` level tests will actually pass without the `Model` class implemented, but I think @selyard has that part done already.

File: app/Controller.cpp (hunk 1 above)
Comment (selyard) [reply app/Controller.cpp]: confirmed as plan, still to be implemented though

File: CMakeLists.txt [hunk 5]
```diff
@@ -15,7 +15,7 @@ if (COVERAGE)
 
     SET(CMAKE_CXX_FLAGS "-g -O0 -fprofile-arcs -ftest-coverage")
     SET(CMAKE_C_FLAGS "-g -O0 -fprofile-arcs -ftest-coverage")
-    SET(CMAKE_EXE_LINKER_FLAGS "-fprofile-arcs -ftest-coverage")
+    SET(CMAKE_EXE_LINKER_FLAGS "-fprofile-arcs -ftest-coverage -lpthread")
```
Comment (danielmohansahu) [inline CMakeLists.txt]:
```suggestion
    SET(CMAKE_EXE_LINKER_FLAGS "-fprofile-arcs -ftest-coverage")
```

File: app/Controller.cpp [hunk 6]
```diff
@@ -80,9 +85,14 @@ void Controller::controlLoop() {
     // update next target time
     next_loop_time += duration;
 
-    //
-    // @TODO groundbreaking robotics goes here
-    //
+    double speed_error, heading_error, command_throttle, command_steering;
+    double dT = 1000/params_->control_frequency;
```
Comment (danielmohansahu) [inline app/Controller.cpp]:
```suggestion
    double dT = 1/params_->control_frequency;
```

**Pull request merged**